Refresh importer project

kellnerd opened the pull request:
In six days the last commit  of the importer project (https://github.com/metabrainz/bookbrainz-utils/commit/2a655229445284b7f84155fa09a8170ca91d5ef9) will have its fourth anniversary.
Time to celebrate this with a huge PR which refreshes almost the entire infrastructure :tada:

## Problems and Solutions

What initially started with a lot of dependency updates and migrations from flow types to TypeScript and from CommonJS modules (CJS) to ES modules (ESM), quickly turned into massive changes to get the project running at all.

There were two aspects which made running the project unsuccessful and extremely cumbersome:
1. **The code itself** contained lots of boilerplate where I'm doubting whether this was working at all back in 2018, for example the [`asyncCluster`](https://github.com/kellnerd/bookbrainz-utils/tree/refresh-importer/importer/src/asyncCluster) for multi-threading and the [queue wrapper](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/queue/index.ts). 
   In the end I've rewritten the queue logic from scratch and did not use `asyncCluster` in my new CLI implementations.
2. **Transpiling** the whole codebase after each coding step before running it is a pain, especially while debugging. I thought I could simply solve that by using `babel-node` to do both steps at once, but unfortunately both `babel-node` and `ts-node` support for ESM has been broken with Node.js v20 and the projects looked more or less stale.
   Since the whole previous setup was effectively unusable, I decided to give [Deno](https://deno.land/) a try. It has native TS and ESM support and the Node.js compatibility is improving with each release. Luckily it worked after I did a few minor changes which are mostly compatible with the previous Node.js/Babel setup. We only have to find a solution to transpile `.ts` import specifiers to `.js` if we ever want to use the previous setup again.

## Main Features

- More reliable reimplementation of the import queue: [src/queue.ts](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/queue.ts)
- Type definitions for [parsed/queued entities](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/parser.ts) and [entity validators](https://github.com/kellnerd/bookbrainz-utils/tree/refresh-importer/importer/src/consumer/validators) (including the migration from the old entity type names Creator and Publication to the current ones)
- New OpenLibrary Import producer CLI: [src/openLibrary/import.ts](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/openLibrary/import.ts)
- New BookBrainz Import Queue (BBIQ) management CLI (fka consumer): [src/bbiq.ts](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/bbiq.ts)

I've tried my best to document the basic usage of the two new CLIs in the [README](https://github.com/kellnerd/bookbrainz-utils/tree/refresh-importer/importer#readme) and through their `--help` flags without wasting too much time for details which might change soon.
For example I'm considering to move the RabbitMQ configuration into the config.json file (once I'm rewriting that part in TS) instead of passing it as a command line argument.
Let me know if I have missed something obvious.

P.S. Currently I'm not sure whether the type definitions for the parsed entities will also be valid and useful for bb-site and whether they should be moved into bb-data. We will better see this later.
(The types for the validators will definitely be moved together with their code, as per [my proposal](https://github.com/kellnerd/gsoc-2023#outdated-and-duplicated-code).)

## How to Review This Beast

This huge branch can best be reviewed commit by commit, I will specifically address important and irrelevant commits now:
- Most important are the new files (with a 2023 copyright header only), the files which I already mentioned above and "feat" commits of course.
- Generally the latest commits are more interesting than the earlier clean-up commits of code which has partially been rewritten or made obsolete since.
- https://github.com/kellnerd/bookbrainz-utils/commit/edada0eb7c60852af5211b6195c4d94491f0d5b2 What should be done with babel-plugin-lodash?
- https://github.com/kellnerd/bookbrainz-utils/commit/27a615613ac6e56f67ce8a8a8548d32c95e55d23 was a lot of work, but I later had to abandon winston, so it is less relevant
- https://github.com/kellnerd/bookbrainz-utils/commit/9db6c2acc33517e40c2ddf250c430ce3de079a56 was the result of a long, unsuccessful debugging session and was reverted later, I have dropped those commits now (I avoided rebasing my original branch in case you were following along and since the branch name `update-deps` was no longer accurate)
- https://github.com/kellnerd/bookbrainz-utils/commit/5c369bfaa57b0128c47aafb8f600ce21b0ead351 is probably the most important gotcha (winston + amqplib = hanging process)

## Overview of Clean-up Tasks

- [x] Get rid of the bluebird promises library
- [x] Upgrade rotten dev dependencies babel et al
- [x] Upgrade vulnerable packages
- [x] Fully migrate from flow to TypeScript
- [x] Migrate from CJS to ESM
- [x] Fix now invalid call signatures and imports
- [x] Use `.ts` and `node:` imports for **Deno** compatibility
  - [x] Avoid using a GitHub package specifier in bb-data's package.json: https://github.com/metabrainz/bookbrainz-data-js/commit/8077fd1ec9b171b528d5a638b992003d9541dcee
- [x] Stop using the `async` npm package
- [ ] Drop useless [error wrappers](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/helpers/errors.ts) (WIP)
- [ ] Delete old [OL producer CLI](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/openLibrary/producer/index.js) (replaced by [new producer CLI](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/openLibrary/import.ts))
- [ ] Delete old [exploration tool](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/openLibrary/exploration) (used to collect statistics about available JSON attributes in OL dumps) and its [old queue dependency](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/queue/index.ts)
- [ ] Delete [`asyncCluster`](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/asyncCluster) (currently no replacement) and uninstall `async` npm package
- [ ] Move [general parser types](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/parser.ts) into a subfolder (which one / how should it be named?)
- [ ] Move [queue module](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/queue.ts) into a subfolder (when the old queue is gone)
- [ ] Improve folder structure, e.g. flatten producer directory
- [ ] Convert [identifier mapping](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/helpers/mapping.ts) into an enum?
- [ ] Update or delete obsolete [README for the config file](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/config/README.md) once the implementation is stable?
- [ ] Update or delete obsolete [package.json scripts](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/package.json) once the implementation is stable

For updates, see [the comment below](https://github.com/metabrainz/bookbrainz-utils/pull/40#issuecomment-1703889451).

MonkeyDo commented on importer/src/queue.ts:
```diff
@@ -0,0 +1,158 @@
+			// Acknowledge all consumed messages and put problematic messages into a separate queue (if configured).
+			// We do this to avoid a requeue/redelivery loop which occurs if the consumer never accepts the entity.
+			this.channel.ack(message);
+			if (!success && this.failureQueueName) {
```
Would it make sense to log errors here if `!success`, regardless of the presence of a failure queue?

kellnerd replied:
Probably not, at this point we only know that the consumer failed but not how. Usually the reason has already been logged by the consumer callback at this point.
We could additionally log them here to be sure we won't miss any discarded message, but honestly the code is already cluttered with too many logger statements for my taste. Most errors are currently logged twice at least, although I had already removed code which first logged an error and then threw it..

MonkeyDo commented on importer/README.md:
````diff
@@ -1,2 +1,94 @@
+Create and run a Docker container with a RabbitMQ server and expose the relevant ports locally:
+
+```sh
+docker run --hostname rabbit --name rabbit-mq -p 127.0.0.1:5672:5672 -p 127.0.0.1:15672:15672 rabbitmq:3
````
I found that I needed to use a specific docker image to get the RMQ management web UI out of the box (as per the [docker image readme](https://github.com/docker-library/docs/tree/master/rabbitmq#management-plugin)).
I would also suggest running the container detached , but that's a detail.
```suggestion
docker run -d --hostname rabbit --name rabbit-mq -p 127.0.0.1:5672:5672 -p 127.0.0.1:15672:15672 rabbitmq:3-management
```

kellnerd replied:
That's actually a good idea, I've even setup the ports for that, but I've always been using the minimal image so far.

MonkeyDo commented:
A few more comments:

> For example I'm considering to move the RabbitMQ configuration into the config.json file (once I'm rewriting that part in TS) instead of passing it as a command line argument.

I think that makes sense. Having DB config in one place and RMQ config in another is confusing. But if you want to keep it, you could always use config as default and use the CLI flags as overrides.
I think that could allow us some more flexibility once we are in production in a docker container (where changing the config that eventually gets written to the config file [[like we have for the website](https://github.com/metabrainz/bookbrainz-site/blob/master/config/config.json.ctmpl)] could be more cumbersome).
——————————

I know we discussed it before, but here it is written down here: I think you should completely ditch the asyncCluster.
No ifs or buts, we can always revisit the code at a later date if the tool gets a lot of regular use (dubious) or if we can definitely justify the potential speed gains compared to dev time.
Machine time is cheap, human brain time less so, I'd rather your focus be on more critical aspects :)

kellnerd commented:
Thanks for going through all of this and for doing an independent test!

I've now also rewritten the configuration file loader (f859f8e) and added support for loading the RabbitMQ connection URL from there to both CLIs (02c233e).

While I was starting my work on the ORM import functions in bb-data, I've found a few more things to be refactored here: ef691cd and 694a28b

> I think you should completely ditch the asyncCluster.

Definitely, the only thing which "stops" me from doing this, is the following:

> Delete old [exploration tool](https://github.com/kellnerd/bookbrainz-utils/blob/refresh-importer/importer/src/openLibrary/exploration) (used to collect statistics about available JSON attributes in OL dumps)

I haven't ported this so far, but if we should ever need this again, I could still recover it from the git history. So I could delete that tool and all its dependencies now (including the old queue wrapper and `asyncCluster`), if you agree.

> The one issue I found was when stopping the consumer manually before the end of the messages list, the process seems to stay open. Not sure if it's a product of my terminal or what, can you confirm?

I will have a look at this later, need to run off now.

kellnerd commented:
> The one issue I found was when stopping the consumer manually before the end of the messages list, the process seems to stay open. Not sure if it's a product of my terminal or what, can you confirm?

I could reproduce this today, it should be fixed by 0d7e500 (and a0f732c should ensure that the process always exits cleanly).
Recently I've been mostly testing with a single row dump to keep log output readable, so it did not happen for me... There have definitely been too many refactoring steps since I had last tested the early cancelling behaviour :innocent:

Now that I'm waiting (up to 2 seconds) for all consumed messages to be actually processed by the callback before I'm closing the queue, `this.channel.ack(message)` should no longer fail.
Instead of the error the output should now look like this:

```
^CConsumer has been terminated
7/10 messages have been processed
3 pending messages still have to be acknowledged before closing...
[TRANSACTION::Author] Error: Error during link import table creation - error: insert into "bookbrainz"."link_import" ("import_id", "import_metadata", "last_edited", "origin_id", "origin_source_id") values ($1, $2, $3, $4, $5) - invalid input syntax for type integer: "{"id":57}"
[CONSUMER::0] Error occurred during DB transaction. :: Error: Error during link import table creation - error: insert into "bookbrainz"."link_import" ("import_id", "import_metadata", "last_edited", "origin_id", "origin_source_id") values ($1, $2, $3, $4, $5) - invalid input syntax for type integer: "{"id":57}" [retry, 0 attempts left]
No more attempts left, giving up
Force-closing, 2 messages are still pending...
Import queue has been force-closed
```

Admittedly I do not understand why there are often "consumed" messages which are never acknowledged, no matter how large I choose the grace period.
Since there was never an attempt to acknowledge them, they cause no error and they are available for consumption again when the consumer is restarted. So that's ok, I guess :shrug:

MonkeyDo reviewed (approved):
Thank you for working on the improvements I suggested and for answering my questions!

You make reviewing large PRs a whole lot easier, with good explanations and clean commits with good descriptions.
Such a treat !

In any case, I think this is ready to merge. We've got a working tool, and definitely a whole lot cleaner than the starting point.

kellnerd commented:
## Remaining Clean-up Tasks

Currently happening in [a new branch](https://github.com/kellnerd/bookbrainz-utils/commits/orm-validators), I might open a new PR with these changes soon. (Unfortunately there are also a few changes in this branch which depend on a [WIP version of bb-data](https://github.com/kellnerd/bookbrainz-data-js/commits/orm-validators), so we might have to wait for these to be merged first.)

- [ ] Drop useless [error wrappers](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/helpers/errors.ts) (WIP)
- [x] Delete old [OL producer CLI](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/openLibrary/producer/index.js) (replaced by [new producer CLI](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/openLibrary/import.ts))
- [x] Delete old [exploration tool](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/openLibrary/exploration) (used to collect statistics about available JSON attributes in OL dumps) and its [old queue dependency](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/queue/index.ts)
- [x] Delete [`asyncCluster`](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/asyncCluster) (currently no replacement) and uninstall `async` npm package
- [x] Move [general parser types](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/parser.ts) into ~a subfolder~ bb-data
- [ ] Move [queue module](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/queue.ts) into a subfolder (when the old queue is gone)
- [ ] Improve folder structure, e.g. flatten producer directory
- [ ] Convert [identifier mapping](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/src/helpers/mapping.ts) into an enum?
- [ ] Update or delete obsolete [README for the config file](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/config/README.md) once the implementation is stable?
- [x] Update or delete obsolete [package.json scripts](https://github.com/kellnerd/bookbrainz-utils/blob/11962796822a3dbdc605c7acd70834a9a2cec6b0/importer/package.json) once the implementation is stable
